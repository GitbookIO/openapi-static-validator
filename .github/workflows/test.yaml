on: [push]
name: Test
jobs:
    test:
        name: Test
        runs-on: ubuntu-latest
        steps:
            - uses: actions/checkout@v3
            - uses: oven-sh/setup-bun@v1
            - run: bun install
            - run: bun run test
