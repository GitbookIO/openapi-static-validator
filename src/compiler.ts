import escodegen from 'escodegen';
import { namedTypes, builders } from 'ast-types';
import { RequestErrorClass, ValidationErrorClass, ValidationErrorIdentifier } from './error';
import { OpenAPIRef, OpenAPISpec } from './types';
import { compileValueSchema } from './compileValueSchema';
import { hash } from './hash';
import { compileValidateRequest } from './compileValidateRequest';

/**
 * Compiler for OpenAPI specs.
 */
export class Compiler {
    private input: OpenAPISpec;

    /** Map of hash to the identifier */
    private hashes: Map<string, string> = new Map();
    /** Map of objects to the identifier */
    private objectHashes: WeakMap<object, string> = new WeakMap();

    /** Counter to get a new identifier */
    private declarationCounter: number = 0;

    /** Map of identifiers defined globally */
    private globalDeclarations: (
        | namedTypes.FunctionDeclaration
        | namedTypes.ClassDeclaration
        | namedTypes.VariableDeclaration
        | namedTypes.ExportNamedDeclaration
    )[] = [
        builders.exportNamedDeclaration(RequestErrorClass),
        builders.exportNamedDeclaration(ValidationErrorClass),
    ];

    /** Map of hashes already processed */
    private processedHashes: Set<string> = new Set();

    constructor(input: OpenAPISpec = {}) {
        this.input = input;
    }

    /**
     * Define a global identifier with a nickname.
     */
    public declareGlobally(
        declaration:
            | namedTypes.FunctionDeclaration
            | namedTypes.ClassDeclaration
            | namedTypes.VariableDeclaration,
    ) {
        this.globalDeclarations.push(declaration);
    }

    /**
     * Declare something globally basded on an input and return an identifier.
     */
    public declareForInput(
        input: any,
        gen: (
            id: namedTypes.Identifier,
        ) =>
            | namedTypes.FunctionDeclaration
            | namedTypes.ClassDeclaration
            | namedTypes.VariableDeclaration,
    ): namedTypes.Identifier {
        const hash = this.hashInput(input);
        const identifier = builders.identifier(hash);

        if (!this.processedHashes.has(hash)) {
            this.processedHashes.add(hash);
            const fn = gen(identifier);
            this.declareGlobally(fn);
        }

        return identifier;
    }

    /**
     * Define a function to validate an input.
     */
    public declareValidationFunction(
        input: object,
        gen: (args: {
            /** Identifier for the value argument being passed to the function */
            value: namedTypes.Identifier;
            /** Identifier for the path argument being passed to the function */
            path: namedTypes.Identifier;
            /** Identifier for the context argument being passed to the function */
            context: namedTypes.Identifier;
            /** Generate an error */
            error: (message: string) => namedTypes.NewExpression;
        }) => namedTypes.BlockStatement['body'],
    ) {
        const pathIdentifier = builders.identifier('path');
        const valueIdentifier = builders.identifier('value');
        const contextIdentifier = builders.identifier('context');

        const error = (message: string) => {
            return builders.newExpression(ValidationErrorIdentifier, [
                pathIdentifier,
                builders.literal(message),
            ]);
        };

        return this.declareForInput(input, (id) => {
            return builders.functionDeclaration(
                id,
                [pathIdentifier, valueIdentifier, contextIdentifier],
                builders.blockStatement(
                    gen({
                        value: valueIdentifier,
                        path: pathIdentifier,
                        context: contextIdentifier,
                        error,
                    }),
                ),
            );
        });
    }

    /**
     * Build the AST from the entire spec.
     */
    public indexAllComponents() {
        // Index all the schema components.
        const schemas = this.input.components?.schemas ?? {};
        Object.values(schemas).forEach((schema) => {
            compileValueSchema(this, schema);
        });
    }

    /**
     * Return the AST for the program.
     */
    public ast() {
        return builders.program([
            ...compileValidateRequest(this, this.input),
            ...this.globalDeclarations,
        ]);
    }

    /**
     * Generate the JS code for the AST.
     */
    public compile() {
        return escodegen.generate(this.ast(), {
            comment: true,
        });
    }

    /**
     * Hash an object and return an identifier name.
     */
    public hashInput(input: any): string {
        const isObject = typeof input === 'object' && input !== null;

        // Fast track for objects
        if (isObject && this.objectHashes.has(input)) {
            return this.objectHashes.get(input)!;
        }

        const hashValue = hash(input);
        if (this.hashes.has(hashValue)) {
            return this.hashes.get(hashValue)!;
        }

        const name = `obj${this.declarationCounter++}`;
        this.hashes.set(hashValue, name);
        if (isObject) {
            this.objectHashes.set(input, name);
        }

        return name;
    }

    /**
     * Resolve a reference to part of the spec.
     * We only support "#/" type of references.
     */
    public resolveRef(ref: OpenAPIRef) {
        const parts = ref.$ref.split('/').slice(1);

        let value: any = this.input;
        for (const part of parts) {
            value = value[part];
            if (value === undefined) {
                throw new Error(`Could not resolve reference ${ref.$ref}`);
            }
        }

        return value;
    }

    /**
     * Resolve a potential ref to a value.
     */
    public resolveMaybeRef<T>(value: T | OpenAPIRef): T {
        if (typeof value === 'object' && value !== null && '$ref' in value) {
            return this.resolveRef(value);
        }

        return value;
    }
}
