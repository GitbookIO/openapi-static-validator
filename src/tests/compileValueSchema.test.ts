import { describe, expect, test } from 'bun:test';
import { Compiler } from '../compiler';
import { compileValueSchema } from '../compileValueSchema';

describe('Number', () => {
    test('basic', () => {
        const compiler = new Compiler();
        compileValueSchema(compiler, {
            type: 'number',
        });
        expect(compiler.compile()).toMatchSnapshot();
    });
});

describe('Integer', () => {
    test('basic', () => {
        const compiler = new Compiler();
        compileValueSchema(compiler, {
            type: 'integer',
            format: 'int32',
        });
        expect(compiler.compile()).toMatchSnapshot();
    });
});

describe('Nullable', () => {
    test('nullable: true', () => {
        const compiler = new Compiler();
        compileValueSchema(compiler, {
            type: 'number',
            nullable: true,
        });
        expect(compiler.compile()).toMatchSnapshot();
    });
});

describe('String', () => {
    test('with enum', () => {
        const compiler = new Compiler();
        compileValueSchema(compiler, {
            type: 'string',
            enum: ['a', 'b', 'c'],
        });
        expect(compiler.compile()).toMatchSnapshot();
    });

    test('with pattern', () => {
        const compiler = new Compiler();
        compileValueSchema(compiler, {
            type: 'string',
            pattern: '^[a-z]+$',
        });
        expect(compiler.compile()).toMatchSnapshot();
    });

    test('with pattern', () => {
        const compiler = new Compiler();
        compileValueSchema(compiler, {
            type: 'string',
            format: 'uri',
        });
        expect(compiler.compile()).toMatchSnapshot();
    });
});

describe('Objects', () => {
    test('with a required prop', () => {
        const compiler = new Compiler();
        compileValueSchema(compiler, {
            type: 'object',
            properties: {
                foo: {
                    type: 'number',
                },
                bar: {
                    type: 'string',
                },
            },
            required: ['foo'],
        });
        expect(compiler.compile()).toMatchSnapshot();
    });

    test('with a default value', () => {
        const compiler = new Compiler();
        compileValueSchema(compiler, {
            type: 'object',
            properties: {
                foo: {
                    type: 'number',
                    default: 10,
                },
            },
        });
        expect(compiler.compile()).toMatchSnapshot();
    });

    test('as free form object', () => {
        const compiler = new Compiler();
        compileValueSchema(compiler, {
            type: 'object',
        });
        expect(compiler.compile()).toMatchSnapshot();
    });

    test('with additionalProperties: true', () => {
        const compiler = new Compiler();
        compileValueSchema(compiler, {
            type: 'object',
            properties: {
                foo: {
                    type: 'number',
                    default: 10,
                },
            },
            additionalProperties: true,
        });
        expect(compiler.compile()).toMatchSnapshot();
    });

    test('with additionalProperties: {}', () => {
        const compiler = new Compiler();
        compileValueSchema(compiler, {
            type: 'object',
            properties: {
                foo: {
                    type: 'number',
                    default: 10,
                },
            },
            additionalProperties: {
                type: 'string',
            },
        });
        expect(compiler.compile()).toMatchSnapshot();
    });

    test('with minProperties/maxProperties', () => {
        const compiler = new Compiler();
        compileValueSchema(compiler, {
            type: 'object',
            minProperties: 1,
            maxProperties: 10,
        });
        expect(compiler.compile()).toMatchSnapshot();
    });
});

describe('Array', () => {
    test('minItems / maxItems', () => {
        const compiler = new Compiler();
        compileValueSchema(compiler, {
            type: 'array',
            items: {
                type: 'string',
            },
            minItems: 1,
            maxItems: 10,
        });
        expect(compiler.compile()).toMatchSnapshot();
    });

    test('uniqueItems', () => {
        const compiler = new Compiler();
        compileValueSchema(compiler, {
            type: 'array',
            items: {
                type: 'string',
            },
            uniqueItems: true,
        });
        expect(compiler.compile()).toMatchSnapshot();
    });
});

test('anyOf', () => {
    const compiler = new Compiler();
    compileValueSchema(compiler, {
        anyOf: [
            {
                type: 'number',
            },
            {
                type: 'string',
            },
        ],
    });
    expect(compiler.compile()).toMatchSnapshot();
});

test('oneOf', () => {
    const compiler = new Compiler();
    compileValueSchema(compiler, {
        oneOf: [
            {
                type: 'number',
            },
            {
                type: 'string',
            },
        ],
    });
    expect(compiler.compile()).toMatchSnapshot();
});

test('allOf', () => {
    const compiler = new Compiler();
    compileValueSchema(compiler, {
        allOf: [
            {
                type: 'object',
                properties: {
                    a: { type: 'number' },
                },
            },
            {
                type: 'object',
                properties: {
                    b: { type: 'string' },
                },
            },
        ],
    });
    expect(compiler.compile()).toMatchSnapshot();
});
