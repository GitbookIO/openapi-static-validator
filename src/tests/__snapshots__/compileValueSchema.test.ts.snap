// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Number basic 1`] = `
"export function validateRequest(request, context) {
    return new ValidationError([], 'no operation match path');
}
class ValidationError extends Error {
    constructor(path, message) {
        super(message);
        this.path = path;
    }
}
function obj0(path, value, context) {
    if (typeof value !== 'number') {
        return new ValidationError(path, 'Expected a number');
    }
    return value;
}"
`;

exports[`Integer basic 1`] = `
"export function validateRequest(request, context) {
    return new ValidationError([], 'no operation match path');
}
class ValidationError extends Error {
    constructor(path, message) {
        super(message);
        this.path = path;
    }
}
function obj0(path, value, context) {
    if (typeof value !== 'number') {
        return new ValidationError(path, 'Expected a number');
    }
    return value;
}"
`;

exports[`Nullable nullable: true 1`] = `
"export function validateRequest(request, context) {
    return new ValidationError([], 'no operation match path');
}
class ValidationError extends Error {
    constructor(path, message) {
        super(message);
        this.path = path;
    }
}
function obj0(path, value, context) {
    if (value === null) {
        return value;
    }
    if (typeof value !== 'number') {
        return new ValidationError(path, 'Expected a number');
    }
    return value;
}"
`;

exports[`String with enum 1`] = `
"export function validateRequest(request, context) {
    return new ValidationError([], 'no operation match path');
}
class ValidationError extends Error {
    constructor(path, message) {
        super(message);
        this.path = path;
    }
}
function obj0(path, value, context) {
    if (value !== 'a' && value !== 'b' && value !== 'c') {
        return new ValidationError(path, 'Expected one of the enum value');
    }
    return value;
}"
`;

exports[`String with pattern 1`] = `
"export function validateRequest(request, context) {
    return new ValidationError([], 'no operation match path');
}
class ValidationError extends Error {
    constructor(path, message) {
        super(message);
        this.path = path;
    }
}
const obj1 = new RegExp('^[a-z]+$');
function obj0(path, value, context) {
    if (typeof value !== 'string') {
        return new ValidationError(path, 'Expected a string');
    }
    if (!obj1.test(value)) {
        return new ValidationError(path, 'Expected to match the pattern "^[a-z]+$"');
    }
    return value;
}"
`;

exports[`Objects with a required prop 1`] = `
"export function validateRequest(request, context) {
    return new ValidationError([], 'no operation match path');
}
class ValidationError extends Error {
    constructor(path, message) {
        super(message);
        this.path = path;
    }
}
function obj1(path, value, context) {
    if (typeof value !== 'number') {
        return new ValidationError(path, 'Expected a number');
    }
    return value;
}
function obj2(path, value, context) {
    if (typeof value !== 'string') {
        return new ValidationError(path, 'Expected a string');
    }
    return value;
}
function obj0(path, value, context) {
    if (typeof value !== 'object' || value === null) {
        return new ValidationError(path, 'Expected an object');
    }
    const keys = new Set(Object.keys(value));
    const value0 = value['foo'];
    if (value0 !== undefined) {
        const result0 = obj1([
            ...path,
            'foo'
        ], value0, context);
        if (result0 instanceof ValidationError) {
            return result0;
        }
        value['foo'] = result0;
        keys.delete('foo');
    } else {
        return new ValidationError(path, 'Expected "foo" to be defined');
    }
    const value1 = value['bar'];
    if (value1 !== undefined) {
        const result1 = obj2([
            ...path,
            'bar'
        ], value1, context);
        if (result1 instanceof ValidationError) {
            return result1;
        }
        value['bar'] = result1;
        keys.delete('bar');
    }
    if (keys.size > 0) {
        return new ValidationError(path, 'Unexpected properties');
    }
    return value;
}"
`;

exports[`Objects with a default value 1`] = `
"export function validateRequest(request, context) {
    return new ValidationError([], 'no operation match path');
}
class ValidationError extends Error {
    constructor(path, message) {
        super(message);
        this.path = path;
    }
}
function obj1(path, value, context) {
    if (typeof value !== 'number') {
        return new ValidationError(path, 'Expected a number');
    }
    return value;
}
function obj0(path, value, context) {
    if (typeof value !== 'object' || value === null) {
        return new ValidationError(path, 'Expected an object');
    }
    const keys = new Set(Object.keys(value));
    const value0 = value['foo'];
    if (value0 !== undefined) {
        const result0 = obj1([
            ...path,
            'foo'
        ], value0, context);
        if (result0 instanceof ValidationError) {
            return result0;
        }
        value['foo'] = result0;
        keys.delete('foo');
    } else {
        value['foo'] = 10;
    }
    if (keys.size > 0) {
        return new ValidationError(path, 'Unexpected properties');
    }
    return value;
}"
`;

exports[`Objects as free form object 1`] = `
"export function validateRequest(request, context) {
    return new ValidationError([], 'no operation match path');
}
class ValidationError extends Error {
    constructor(path, message) {
        super(message);
        this.path = path;
    }
}
function obj0(path, value, context) {
    if (typeof value !== 'object' || value === null) {
        return new ValidationError(path, 'Expected an object');
    }
    const keys = new Set(Object.keys(value));
    return value;
}"
`;

exports[`Objects with additionalProperties: true 1`] = `
"export function validateRequest(request, context) {
    return new ValidationError([], 'no operation match path');
}
class ValidationError extends Error {
    constructor(path, message) {
        super(message);
        this.path = path;
    }
}
function obj1(path, value, context) {
    if (typeof value !== 'number') {
        return new ValidationError(path, 'Expected a number');
    }
    return value;
}
function obj0(path, value, context) {
    if (typeof value !== 'object' || value === null) {
        return new ValidationError(path, 'Expected an object');
    }
    const keys = new Set(Object.keys(value));
    const value0 = value['foo'];
    if (value0 !== undefined) {
        const result0 = obj1([
            ...path,
            'foo'
        ], value0, context);
        if (result0 instanceof ValidationError) {
            return result0;
        }
        value['foo'] = result0;
        keys.delete('foo');
    } else {
        value['foo'] = 10;
    }
    return value;
}"
`;

exports[`Objects with additionalProperties: {} 1`] = `
"export function validateRequest(request, context) {
    return new ValidationError([], 'no operation match path');
}
class ValidationError extends Error {
    constructor(path, message) {
        super(message);
        this.path = path;
    }
}
function obj1(path, value, context) {
    if (typeof value !== 'number') {
        return new ValidationError(path, 'Expected a number');
    }
    return value;
}
function obj2(path, value, context) {
    if (typeof value !== 'string') {
        return new ValidationError(path, 'Expected a string');
    }
    return value;
}
function obj0(path, value, context) {
    if (typeof value !== 'object' || value === null) {
        return new ValidationError(path, 'Expected an object');
    }
    const keys = new Set(Object.keys(value));
    const value0 = value['foo'];
    if (value0 !== undefined) {
        const result0 = obj1([
            ...path,
            'foo'
        ], value0, context);
        if (result0 instanceof ValidationError) {
            return result0;
        }
        value['foo'] = result0;
        keys.delete('foo');
    } else {
        value['foo'] = 10;
    }
    for (const key of keys) {
        const result = obj2([
            ...path,
            key
        ], value[key], context);
        if (result instanceof ValidationError) {
            return result;
        }
        value[key] = result;
    }
    return value;
}"
`;

exports[`Objects with minProperties/maxProperties 1`] = `
"export function validateRequest(request, context) {
    return new ValidationError([], 'no operation match path');
}
class ValidationError extends Error {
    constructor(path, message) {
        super(message);
        this.path = path;
    }
}
function obj0(path, value, context) {
    if (typeof value !== 'object' || value === null) {
        return new ValidationError(path, 'Expected an object');
    }
    const keys = new Set(Object.keys(value));
    if (keys.size < 1) {
        return new ValidationError(path, 'Expected at least 1 properties');
    }
    if (keys.size > 10) {
        return new ValidationError(path, 'Expected at most 10 properties');
    }
    return value;
}"
`;

exports[`Array minItems / maxItems 1`] = `
"export function validateRequest(request, context) {
    return new ValidationError([], 'no operation match path');
}
class ValidationError extends Error {
    constructor(path, message) {
        super(message);
        this.path = path;
    }
}
function obj1(path, value, context) {
    if (typeof value !== 'string') {
        return new ValidationError(path, 'Expected a string');
    }
    return value;
}
function obj0(path, value, context) {
    if (!Array.isArray(value)) {
        return new ValidationError(path, 'Expected an array');
    }
    if (value.length < 1) {
        return new ValidationError(path, 'Expected at least 1 items');
    }
    if (value.length > 10) {
        return new ValidationError(path, 'Expected at most 10 items');
    }
    for (let i = 0; i < value.length; i++) {
        const itemResult = obj1([
            ...path,
            i
        ], value[i], context);
        if (itemResult instanceof ValidationError) {
            return itemResult;
        }
        value[i] = itemResult;
    }
    return value;
}"
`;

exports[`Array uniqueItems 1`] = `
"export function validateRequest(request, context) {
    return new ValidationError([], 'no operation match path');
}
class ValidationError extends Error {
    constructor(path, message) {
        super(message);
        this.path = path;
    }
}
function obj1(path, value, context) {
    if (typeof value !== 'string') {
        return new ValidationError(path, 'Expected a string');
    }
    return value;
}
function obj0(path, value, context) {
    if (!Array.isArray(value)) {
        return new ValidationError(path, 'Expected an array');
    }
    const valueSet = new Set();
    for (let i = 0; i < value.length; i++) {
        const itemResult = obj1([
            ...path,
            i
        ], value[i], context);
        if (valueSet.has(itemResult)) {
            return new ValidationError(path, 'Expected unique items');
        }
        valueSet.add(itemResult);
        if (itemResult instanceof ValidationError) {
            return itemResult;
        }
        value[i] = itemResult;
    }
    return value;
}"
`;

exports[`anyOf 1`] = `
"export function validateRequest(request, context) {
    return new ValidationError([], 'no operation match path');
}
class ValidationError extends Error {
    constructor(path, message) {
        super(message);
        this.path = path;
    }
}
function obj1(path, value, context) {
    if (typeof value !== 'number') {
        return new ValidationError(path, 'Expected a number');
    }
    return value;
}
function obj2(path, value, context) {
    if (typeof value !== 'string') {
        return new ValidationError(path, 'Expected a string');
    }
    return value;
}
function obj0(path, value, context) {
    const value0 = obj1(path, value, context);
    if (!(value0 instanceof ValidationError)) {
        return value0;
    }
    const value1 = obj2(path, value, context);
    if (!(value1 instanceof ValidationError)) {
        return value1;
    }
    return new ValidationError(path, 'Expected one of the anyOf schemas to match');
}"
`;

exports[`oneOf 1`] = `
"export function validateRequest(request, context) {
    return new ValidationError([], 'no operation match path');
}
class ValidationError extends Error {
    constructor(path, message) {
        super(message);
        this.path = path;
    }
}
function obj1(path, value, context) {
    if (typeof value !== 'number') {
        return new ValidationError(path, 'Expected a number');
    }
    return value;
}
function obj2(path, value, context) {
    if (typeof value !== 'string') {
        return new ValidationError(path, 'Expected a string');
    }
    return value;
}
function obj0(path, value, context) {
    let result;
    const alt0 = obj1(path, value, context);
    if (!(alt0 instanceof ValidationError)) {
        result = alt0;
    }
    const alt1 = obj2(path, value, context);
    if (!(alt1 instanceof ValidationError)) {
        result = alt1;
        if (result !== undefined) {
            return new ValidationError(path, 'Expected to only match one of the schemas');
        }
    }
    return result;
}"
`;

exports[`allOf 1`] = `
"export function validateRequest(request, context) {
    return new ValidationError([], 'no operation match path');
}
class ValidationError extends Error {
    constructor(path, message) {
        super(message);
        this.path = path;
    }
}
function obj2(path, value, context) {
    if (typeof value !== 'number') {
        return new ValidationError(path, 'Expected a number');
    }
    return value;
}
function obj1(path, value, context) {
    if (typeof value !== 'object' || value === null) {
        return new ValidationError(path, 'Expected an object');
    }
    const keys = new Set(Object.keys(value));
    const value0 = value['a'];
    if (value0 !== undefined) {
        const result0 = obj2([
            ...path,
            'a'
        ], value0, context);
        if (result0 instanceof ValidationError) {
            return result0;
        }
        value['a'] = result0;
        keys.delete('a');
    }
    if (keys.size > 0) {
        return new ValidationError(path, 'Unexpected properties');
    }
    return value;
}
function obj4(path, value, context) {
    if (typeof value !== 'string') {
        return new ValidationError(path, 'Expected a string');
    }
    return value;
}
function obj3(path, value, context) {
    if (typeof value !== 'object' || value === null) {
        return new ValidationError(path, 'Expected an object');
    }
    const keys = new Set(Object.keys(value));
    const value0 = value['b'];
    if (value0 !== undefined) {
        const result0 = obj4([
            ...path,
            'b'
        ], value0, context);
        if (result0 instanceof ValidationError) {
            return result0;
        }
        value['b'] = result0;
        keys.delete('b');
    }
    if (keys.size > 0) {
        return new ValidationError(path, 'Unexpected properties');
    }
    return value;
}
function obj0(path, value, context) {
    let result = value;
    result = obj1(path, result, context);
    if (result instanceof ValidationError) {
        return result;
    }
    result = obj3(path, result, context);
    if (result instanceof ValidationError) {
        return result;
    }
    return result;
}"
`;
